pkg/util/aad/aad.go

```diff
@@ -39,44 +74,28 @@ func GetToken(ctx context.Context, log *logrus.Entry, clientID, clientSecret, te
 	// during credentials rotation this can take time to propagate
 	timeoutCtx, cancel := context.WithTimeout(ctx, 5*time.Minute)
 	defer cancel()
-
+	done := false
 	// NOTE: Do not override err with the error returned by
 	// wait.PollImmediateUntil. Doing this will not propagate the latest error
 	// to the user in case when wait exceeds the timeout
-	_ = wait.PollImmediateUntil(10*time.Second, func() (bool, error) {
-		var done bool
-		done, err = authorizer.RefreshWithContext(ctx, log)
-		if err != nil {
-			err = api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidServicePrincipalCredentials, "properties.servicePrincipalProfile", "The provided service principal credentials are invalid.")
-		}
-		if !done || err != nil {
-			return false, err
-		}
-
-		p := &jwt.Parser{}
-		claims := jwt.MapClaims{}
-		_, _, err = p.ParseUnverified(authorizer.OAuthToken(), claims)
-		if err != nil {
-			err = api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidServicePrincipalToken, "properties.servicePrincipalProfile", "The provided service principal generated an invalid token.")
-			return false, err
-		}
 
-		for _, claim := range []string{"altsecid", "oid", "puid"} {
-			if _, found := claims[claim]; found {
-				return true, nil
-			}
+	_ = wait.PollImmediateUntil(10*time.Second, func() (bool, error) {
+		err = refreshContext(&authorizer, &ctx, &done, log)
+		if done {
+			done, err = parseAuthToken(&authorizer, log)
 		}
-
-		// populate err with a user-facing error that will be visible if we're
-		// not successful.
-		log.Info("token does not contain the required claims")
-		err = api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidServicePrincipalClaims, "properties.servicePrincipalProfile", "The provided service principal does not give an access token with at least one of the claims 'altsecid', 'oid' or 'puid'.")
-
 		return false, nil
```

Comment:
actually this should probably be a single function, and it would make it slightly easier to test